man/ulp.1
.\" libpulp - User-space Livepatching Library
.\"
.\" Copyright (C) 2021 SUSE Software Solutions GmbH
.\"
.\" This file is part of libpulp.
.\"
.\" libpulp is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU Lesser General Public
.\" License as published by the Free Software Foundation; either
.\" version 2.1 of the License, or (at your option) any later version.
.\"
.\" libpulp is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" Lesser General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with libpulp.  If not, see <http://www.gnu.org/licenses/>.

.TH ULP 1 "" "" "Libpulp Tools"
.SH NAME
ulp \- Userspace Livepatching Tool
.SH SYNOPSIS
.B ulp command
[SWITCHES]...
[ARGUMENTS]
.I file
.SH INTRODUCTION
.B ulp
consists of a set of tools used in the userspace livepatching process. Those
include querying patch status, libpulp messages, applying and removing patches,
and checking for livepatch capabilities within a process.
.TP
Commands currently supported by ulp are:
.TP
.B patches:
list running programs that have livepatching capabilities loaded, as well the
libraries that are livepatchable and the list of loaded patches.
.TP
.B check:
check if a certain livepatch (.so) is applied to
a running process.
.TP
.B dump:
print the content of a livepatch (.so) in human-readable form.
.TP
.B packer:
compile the livepatch descriptor (.dsc file) into a livepatch
(.ulp file). The later can be used to apply a livepatch to a running process
using trigger.
.TP
.B trigger:
apply or remove a list of livepatches to a list of running processes.
.TP
.B post:
apply any post-processing needed to livepatch containers (.so files).
.TP
.B messages:
querry the libpulp internal message queue in a running process. This can be
used to diagnose livepaching errors that occured internally in the target
process.
.TP
.TP
Detailed descriptions about how these commands work are provided in the next
sections.

.\"-------------------------------------------

.SH PATCHES

.TP
.SH NAME
ulp patches \- Show active live patches and libraries
.TP
.SH SYNOPSIS
.B ulp patches
.BR -p wildcard
.TP
.SH DESCRIPTION
.B ulp patches show which libraries are livepatchable and which livepatches
are loaded in process that matches either
.I wildcard
as name or with pid =
.I wildcard.
.TP
.SH OPTIONS
.TP
.B -p
.I wildcard
Show processes which name matches the wildcard, or with pid = wildcard.
.TP
.SH EXIT STATUS
.TP
.B ulp patches
exits 0 on success, anything else on error.

.\"-------------------------------------------

.SH CHECK

.TP
.SH NAME
ulp check \- Apply a live patch
.TP
.SH SYNOPSIS
.B ulp check
[OPTION]...
.BR -p pid
.I file
.TP
.SH DESCRIPTION
.B ulp check
attaches to the target process specified by
.I pid
and checks whether the live patch described by the metadata
.I file
that it takes as argument has already been applied or not.
.PP
Only relevant processes can be inspected with
.BR ulp check .
More specifically, the following conditions must be met:
.TP
.B Attaching to the Target Process
Libpulp uses ptrace to attach to target processes, thus only the owner of a
process (or root) can apply a live patch to it. Moreover, on systems that have
Linux Security Modules (LSM) enabled, some extra convincing might be required
(see Ptrace access mode checking in
.IR ptrace (2)).
When
.B ulp check
is unable to attach to the target process, it exits in error.
.TP
.B Runtime support
Applying a live patch requires functions from libpulp.so, which must have been
loaded into the address space of the target process. Typically, this is
accomplished with LD_PRELOAD (see
.IR libpulp (7)).
.TP
.SH FATAL ERRORS
If a problem happens during the execution of functions from libpulp.so, the
target process might end up in an inconsistent state. When that happens
.B ulp check
exits with error code
.BR -1 ,
and the user is advised to kill the process.
.TP
.SH OPTIONS
.TP
.B -q, --quiet
Do not produce any output.
.TP
.B -v, --verbose
Produce verbose output.
.TP
.B -?, --help
Display a lengthy usage message.
.TP
.B -usage
Display a brief usage message.
.TP
.B -V --version
Print program version and exit.
.TP
.SH BUGS
.TP
.B ulp check
also returns -1 on non-fatal errors.
.TP
.SH EXIT STATUS
.TP
.B ulp check
.\" XXX: ulp_check curretly returns -1 on non-fatal errors.
exits 0 on success, 1 on error, and -1 on fatal errors. A fatal error is an
indication that the target process should be killed, because it was probably
left in an inconsistent state.

.\"-------------------------------------------

.SH DUMP
.TP
.SH NAME
ulp dump \- Prints metadata information in human-readable format
.TP
.SH SYNOPSIS
.TP
.B ulp dump
.I file
.TP
.SH DESCRIPTION
.TP
.B ulp dump
parses the metadata
.I file
that it takes as argument, which is in binary format, then prints its content
to the standard output in human-readable format.
.PP
.TP
.SH EXIT STATUS
.TP
.B ulp dump
exits 0 on success and 1 on error.

.\"-------------------------------------------

.SH PACKER
.TP
.SH NAME
ulp packer \- Create live patch metadata
.TP
.SH SYNOPSIS
.B ulp packer
[OPTION]...
.I file_wildcard
.TP
.SH DESCRIPTION
.B ulp packer
creates a live patch metadata file based on the live patch description
.I file
that it takes as argument.
After parsing the description file,
.B ulp packer
validates that the target library and live patch objects referred to exist,
then produces the metadata file required by live patching tools, such as
.BR ulp
.BR trigger (1)
and
.BR ulp
.BR check (1).
.PP
The syntax of the description file is described in
.IR libpulp (7).
.PP
By default, the output is written to stdout, but it can be optionally
redirected to a specified file. See OPTIONS below.
.TP
.SH OPTIONS
.TP
.B -o, --output=FILENAME
Instead of printing the results to the standard output, write them to FILENAME.
.TP
.B -p, --process=WILDCARD
Patch any process which name matches the WILDCARD. A single PID is also
supported by passing WILDCARD=PID of the target process.

.B -l, --livepatch=FILENAME
Instead of getting the path to the live patch object from the description file,
use FILENAME.
.TP
.B -t, --target=FILENAME
Instead of getting the path to the target library from the description file,
use FILENAME.
.TP
.B -q, --quiet
Do not produce any output.
.TP
.B -v, --verbose
Produce verbose output.
.TP
.B -?, --help
Display a lengthy usage message.
.TP
.B -usage
Display a brief usage message.
.TP
.B -V --version
Print program version and exit.
.TP
.B -R --root=PREFIX
Append PREFIX to the path to the livepatch .so file when it is send to the
target process. This is useful if ulp is running inside a chroot.
.TP
.SH EXIT STATUS
.B ulp packer
exits 0 on success and 1 on error.

.\"-------------------------------------------

.SH TRIGGER
.TP
.SH NAME
ulp trigger \- Apply a live patch
.TP
.SH SYNOPSIS
.B ulp trigger
[OPTION]...
.BR -p
.I process_wildcard
.I file_wildcard
.TP
.SH DESCRIPTION
.B ulp trigger
attaches to the target process specified by a set of process in
.I process_wildcard
or a single process with pid equals to
.I process_wildcard
and applies a set of livepatches described by the metadata
.I file_wildcard
that it takes as argument.
After parsing the metadata file, several checks are performed to verify that
the target process can receive the specified live patch:
.TP
.B Target Library
A live patch replaces functions belonging to a shared library, thus,
.B ulp trigger
searches the memory space of the target process for its presence. When the
library is not present,
.B ulp trigger
exits in error.
.TP
.B Replacement functions
The metadata file contains a list of replacement functions, which must be
present in the live patch object (DSO). If all functions are present, the live
patching operation can proceed, otherwise
.B ulp trigger
exits in error.
.TP
.B Attaching to the Target Process
Libpulp uses ptrace to attach to target processes, thus only the owner of a
process (or root) can apply a live patch to it. Moreover, on systems that have
Linux Security Modules (LSM) enabled, some extra convincing might be required
(see Ptrace access mode checking in
.IR ptrace (2)).
When
.B ulp trigger
is unable to attach to the target process, it exits in error.
.TP
.B Runtime support
Applying a live patch requires functions from libpulp.so, which must have been
loaded into the address space of the target process. Typically, this is
accomplished with LD_PRELOAD (see
.IR libpulp (7)).
.TP
.B Forward progress
After attaching to the target process with ptrace, Libpulp calls functions from
libpulp.so. The execution of these functions happens from the context of a
signal handler, thus
.I AS-Unsafe
functions are not allowed (see
.IR attributes (7)).
However, Libpulp requires the use of
.IR malloc (3),
.IR dlopen (3)
and
.IR dlsym (3),
which are all
.IR AS-Unsafe .
In order to avoid deadlocks, libpulp.so checks that these functions are not in
execution anywhere in the target process, before starting the live patching
operation.
.TP
.SH FATAL ERRORS
If a problem happens after Libpulp started replacing functions from the target
process, the process might end up in an inconsistent state. When that happens
.B ulp trigger
exits with error code
.BR -1 ,
and the user is advised to kill the process.
.TP
.SH OPTIONS
.TP
.B -r, --retries=N
To guarantee
.BR "Forward Progress" ,
Libpulp first checks whether trying to apply a live patch would cause a
deadlock in the target process, or if it would be safe to do so. By default,
.B ulp trigger
performs this check a single time and exits in error if the check fails.
However, the state of the relevant locks usually changes very quickly, thus,
there is a high chance that trying again after a brief moment would allow the
live patching operation to succeed without risk of deadlock. This option tells
.B ulp trigger
to try again
.I N
times.
.TP
.B -c, --check-stack
Before applying the live patch to the target process, unwind the stacks of all
threads and make sure that none of them have library calls in execution. If any
thread is within the target library,
.B ulp trigger
aborts the live patching operation; on the other hand, if no threads are within
the target library, the live patch can be applied with additional consistency
guarantees.
.TP
.B --revert-all=LIB
Before applying the live patch to the target process, revert all livepatches
applied to the library LIB. If LIB=target, then all patches to the target
library of the livepatch will be removed.
.TP
.B --disable-summarization
Disable output summarization. This avoids suppression of output 'irrelevant output'
with regard to skipped livepatches.
.TP
.B --timeout N
Wait N seconds for a reply from libpulp. Default is 100s. In cases where the
system is busy running multiple tasks it may be worth increasing this number,
once ulp will bail out to not hang a system update.
.TP
.B --recursive
Look for livepatches recursively when a wildcard is passed.
.TP
.B -q, --quiet
Do not produce any output.
.TP
.B -v, --verbose
Produce verbose output.
.TP
.B -?, --help
Display a lengthy usage message.
.TP
.B -usage
Display a brief usage message.
.TP
.B -V --version
Print program version and exit.
.TP
.SH EXIT STATUS
.B ulp trigger
exits 0 on success, 1 on error, and -1 on fatal errors. A fatal error is an
indication that the target process should be killed, because it was probably
left in an inconsistent state.

.\"-------------------------------------------

.SH POST
.TP
.SH NAME
ulp post \- Post-process live patchable libraries
.TP
.SH SYNOPSIS
.B ulp post
.I file
.TP
.SH DESCRIPTION
.B ulp post
opens the library
.I file
passed as argument and replaces one-byte nops with multi-byte nop instructions
at patchable function entries (see
.IR gcc (1)).
.TP
.SH EXIT STATUS
.B ulp post
exits 0 on success, and 1 on error.

.\"-------------------------------------------

.SH REVERSE
.TP
.SH NAME
ulp reverse \- Create live patch metadata
.TP
.SH SYNOPSIS
.TP
.B ulp reverse
.I file
.TP
.SH DESCRIPTION
.TP
.B ulp reverse
creates a live patch metadata used to revert the effects of the metadata
.I file
that it takes as argument.
Live patch reversal does not require a live patch object file, because it does
not add new replacement functions; rather, it causes the reverse-patched
process to fallback to the functions that had been previously replaced. These
functions are already present in the memory space of the target process.
.PP
.TP
.SH EXIT STATUS
.TP
.B ulp reverse
exits 0 on success and a positive integer on error.

.\"-------------------------------------------

.SH MESSAGES
.TP
.SH NAME
ulp messages \- Querry internal messages from libpulp
.TP
.SH SYNOPSIS
.TP
.B ulp messages
-p
.I pid
.TP
.SH DESCRIPTION
.TP
.B ulp messages
print all internal messages from libpulp message queue in process running
with pid =
.I pid
Those messages are useful to debug any problem that may happens when a livepatch
is applied.
.TP
Messages are output to stdout.
.PP
.TP
.SH EXIT STATUS
.TP
.B ulp messages
exits 0 on success, anything else on error.

.\"-------------------------------------------

.SH EXTRACT
.TP
.SH NAME
ulp extract \- Extract the relevant content from a livepatchable library
.TP
.SH SYNOPSIS
.TP
.B ulp extract
.I livepatchable_library
-o
.I output_file
.TP
.SH DESCRIPTION
.TP
Extract the relevant content from
.I livepatchable_library
such as the library name, buildid, and symbols in the library; and write those
informations into a JSON file specified by
.I output_file.
.PP
.TP
.SH EXIT STATUS
.TP
exits 0 on success, anything else on error.

.\"-------------------------------------------

.SH SET_PATCHABLE
.TP
.SH NAME
ulp set_patchable \- Enable/disable livepatching on target process
.TP
.SH SYNOPSIS
.TP
.B ulp set_livepatchable -p process
.I enable/disable
.TP
.SH DESCRIPTION
.TP
Enable or disable the livepatching capability of process specified by
-p
.I WILDCARD.
If WILDCARD is a number, then it is assumed to be the PID of the process.
The program will output which processes were modified.
.PP
.TP
.SH EXIT STATUS
.TP
Always returns 0.


tools/ulp.c
/*
 *  libpulp - User-space Livepatching Library
 *
 *  Copyright (C) 2020-2021 SUSE Software Solutions GmbH
 *
 *  This file is part of libpulp.
 *
 *  libpulp is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  libpulp is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with libpulp.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <argp.h>
#include <dirent.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>

#include "arguments.h"
#include "check.h"
#include "config.h"
#include "dump.h"
#include "extract.h"
#include "introspection.h"
#include "livepatchable.h"
#include "messages.h"
#include "packer.h"
#include "patches.h"
#include "post.h"
#include "set_patchable.h"
#include "trigger.h"

#include <unistd.h>

#ifdef ENABLE_AFL
#define AFL_INIT_SET0(_p) \
  do { \
    argv = afl_init_argv(&argc); \
    argv[0] = (_p); \
    if (!argc) \
      argc = 1; \
  } \
  while (0)

#define MAX_CMDLINE_LEN 100000
#define MAX_CMDLINE_PAR 1000

__attribute__((unused)) static char **
afl_init_argv(int *argc)
{

  static char in_buf[MAX_CMDLINE_LEN];
  static char *ret[MAX_CMDLINE_PAR];

  char *ptr = in_buf;
  int rc = 0;

  if (read(0, in_buf, MAX_CMDLINE_LEN - 2) < 0) {
  }

  while (*ptr) {

    ret[rc] = ptr;
    if (ret[rc][0] == 0x02 && !ret[rc][1])
      ret[rc]++;
    rc++;

    while (*ptr)
      ptr++;
    ptr++;
  }

  *argc = rc;

  return ret;
}

#undef MAX_CMDLINE_LEN
#undef MAX_CMDLINE_PAR
#endif /* ENABLE_AFL  */

static error_t parser(int, char *, struct argp_state *);
static bool check_color_available(void);

/* These variables are used by ARGP.  */

const char *argp_program_version = PACKAGE_STRING;

static const char args_doc[] = "COMMAND [ARG1 ARG2 ...]";

/* clang-format off */
static const char doc[] =
"ulp: Userspace Live Patch tool.\n"
"\n"
" This tool executes a COMMAND passed in the argument list.\n"
" Possible COMMANDs are:\n"
"\n"
"   patches                   List active patches.\n"
"   check                     Check if patch in ARG1 is applied on process\n"
"                             with -p PID.\n"
"   dump                      Print the content of metadata file on ARG1 in\n"
"                             human readable form.\n"
"   packer                    Creates a livepatch METADATA file based on a\n"
"                             live patch description on ARG1.\n"
"   trigger                   Applies the live patch in ARG1 to the process\n"
"                             with PID\n"
"   post                      Post process patch container (.so file) in ARG1.\n"
"   messages                  Print livepatch information contained in libpulp.\n"
"   livepatchable             Check if .so library in ARG1 is livepatchable.\n"
"   set_patchable             Enable/disable livepatching in process given by -p.\n";

/* clang-format on */

/* switches that don't have a shorthand.  */
#define ULP_OP_REVERT_ALL 256
#define ULP_OP_REVERT 257
#define ULP_OP_COLOR 258
#define ULP_OP_TIMEOUT 259
#define ULP_OP_DISABLE_THREADING 260
#define ULP_OP_RECURSIVE 261
#define ULP_OP_DISABLE_SUMMARIZATION 262
#define ULP_OP_ONLY_LIVEPATCHED 263

static struct argp_option options[] = {
  { 0, 0, 0, 0, "Options:", 0 },
  { "verbose", 'v', 0, 0, "Produce verbose output", 0 },
  { "quiet", 'q', 0, 0, "Don't produce any output", 0 },
  { 0, 0, 0, 0, "patches, check & trigger commands only:", 0 },
  { "process", 'p', "process", 0, "Target process name, wildcard, or PID", 0 },
  { "user", 'u', "user", 0, "User name, wildcard, or UID", 0 },
  { "disable-threading", ULP_OP_DISABLE_THREADING, 0, 0,
    "Do not launch additional threads", 0 },
  { 0, 0, 0, 0, "dump & patches command only:", 0 },
  { "buildid", 'b', 0, 0, "Print the build id", 0 },
  { "only-livepatched", ULP_OP_ONLY_LIVEPATCHED, 0, 0, "Print only processes that were livepatched", 0 },
  { 0, 0, 0, 0, "trigger command only:", 0 },
  { "revert-all", ULP_OP_REVERT_ALL, "LIB", 0,
    "Revert all patches from LIB. If LIB=target, then all patches from the "
    "target library within the passed livepatch will be reverted.",
    0 },
  { "timeout", ULP_OP_TIMEOUT, "t", 0,
    "Set trigger timeout to t seconds (default 200s)", 0 },
  { "disable-summarization", ULP_OP_DISABLE_SUMMARIZATION, 0, 0,
    "Disable trigger ouput summarization", 0 },
  { "recursive", ULP_OP_RECURSIVE, 0, 0, "Search for patches recursively", 0 },
  { "root", 'R', "PREFIX", 0,
    "Append prefix to livepatch path when passing it to target process", 0 },
#if defined ENABLE_STACK_CHECK && ENABLE_STACK_CHECK
  { "check-stack", 'c', 0, 0, "Check the call stack before live patching", 0 },
#endif
  { 0, 0, 0, 0, "trigger & set_patchable commands only:", 0 },
  { "retries", 'r', "N", 0, "Retry N times if process busy", 0 },
  { 0, 0, 0, 0, "trigger & dump commands only:", 0 },
  { "revert", ULP_OP_REVERT, 0, 0,
    "revert livepatch / dump reverse patch info.", 0 },
  { 0, 0, 0, 0, "packer commands only:", 0 },
  { "output", 'o', "FILE", 0, "Write output to FILE", 0 },
  { 0, 0, 0, 0, "packer command only:", 0 },
  { "livepatch", 'l', "LIVEPATCH", 0,
    "Use this livepatch file\nDefaults to the one described in ARG1", 0 },
  { "target", 't', "LIBRARY", 0,
    "Use this target library\nDefaults to the one described in ARG1", 0 },
  { "color", ULP_OP_COLOR, "yes/no/auto", 0, "Enable/disable colored messages", 0 },
  { 0, 0, 0, 0, "extract command only:", 0 },
  { "with-debuginfo", 'd', "DEBUGINFO", 0,
    "Use debuginfo information for symbolextraction", 0 },
  { 0 }
};

static struct argp argp = { options, parser, args_doc, doc, NULL, NULL, NULL };

/* End of variables used by argp.  */

/* Get command string and return an command code for it.
 *
 * @param str - String representation of command.
 * @return code of command.
 */

static command_t
command_from_string(const char *str)
{
  struct entry
  {
    const char *string;
    command_t command;
  };

  static const struct entry entries[] = {
    { "patches", ULP_PATCHES },   { "check", ULP_CHECK },
    { "dump", ULP_DUMP },         { "packer", ULP_PACKER },
    { "trigger", ULP_TRIGGER },   { "post", ULP_POST },
    { "messages", ULP_MESSAGES }, { "livepatchable", ULP_LIVEPATCHABLE },
    { "extract", ULP_EXTRACT },   { "set_patchable", ULP_SET_PATCHABLE },
  };

  size_t i;

  for (i = 0; i < ARRAY_LENGTH(entries); i++) {
    if (!strcmp(entries[i].string, str))
      return entries[i].command;
  }

  return ULP_NONE;
}

/* This function is called when all arguments have been parsed.  */
static void
handle_end_of_arguments(const struct argp_state *state)
{
  const struct arguments *arguments = state->input;
  int path_length;

  if (state->arg_num < 1)
    argp_error(state, "Too few arguments.");

  if (arguments->quiet && arguments->verbose)
    argp_error(state, "You must specify either '-v' or '-q' or none.");

  switch (arguments->command) {
    case ULP_NONE:
      argp_error(state, "Invalid command.");
      break;

    case ULP_PATCHES:
      if (state->arg_num > 1)
        argp_error(state, "Too many arguments.");
      break;

    case ULP_EXTRACT:
    case ULP_PACKER:
    case ULP_CHECK:
      if (state->arg_num < 2)
        argp_error(state, "Too few arguments.");
      break;

    case ULP_TRIGGER:
      if (arguments->library) {
        /* revert-all was passed to trigger.  */
        if (arguments->revert)
          argp_error(state,
                     "--revert and --revert-all can not be used together.");

        if (state->arg_num < 1)
          argp_error(state, "Too few arguments.");

        path_length = strlen(arguments->library);
        if (path_length > ULP_PATH_LEN)
          argp_error(state, "LIB name must be shorter than %d bytes; got %d.",
                     ULP_PATH_LEN, path_length);
      }
      else {
        /* revert-all was not passed to trigger. Metadata file is required.  */
        if (state->arg_num < 2)
          argp_error(state, "Too few arguments.");
      }

      if (state->arg_num >= 2) {
        path_length = strlen(arguments->args[0]);
        if (path_length > ULP_PATH_LEN)
          argp_error(state,
                     "METADATA path must be shorter than %d bytes; got %d.",
                     ULP_PATH_LEN, path_length);
      }
      break;

    /* Currently, DUMP & POST does the same checks.  */
    case ULP_POST:
    case ULP_DUMP:
      if (state->arg_num < 2)
        argp_error(state, "Too few arguments.");

      path_length = strlen(arguments->args[0]);
      if (path_length > ULP_PATH_LEN)
        argp_error(state,
                   "METADATA path must be shorter than %d bytes; got %d.",
                   ULP_PATH_LEN, path_length);
      break;

    case ULP_MESSAGES:
      if (arguments->process_wildcard == 0)
        argp_error(state, "process is mandatory in 'messages' command.");
      break;

    case ULP_SET_PATCHABLE:
      if (state->arg_num < 2)
        argp_error(state, "passing 'enable' or 'disable' in ARG1 is mandatory "
                          "in set_patches.");
      break;

    case ULP_LIVEPATCHABLE:
      if (state->arg_num < 2) {
        argp_error(state, "file is mandatory in 'livepatchable' command.");
      }
  }
}

/* Parse the arguments in command line.  */
static error_t
parser(int key, char *arg, struct argp_state *state)
{
  struct arguments *arguments;

  arguments = state->input;

  switch (key) {
    case 'v':
      arguments->verbose = 1;
      break;
    case 'q':
      arguments->quiet = 1;
      break;
    case 'p':
      arguments->process_wildcard = arg;
      break;
    case 'b':
      arguments->buildid = 1;
      break;
    case 'o':
      arguments->metadata = arg;
      break;
    case 'l':
      arguments->livepatch = arg;
      break;
    case 't':
      arguments->library = arg;
      break;
    case 'r':
      arguments->retries = atoi(arg);
      if (arguments->retries < 1)
        argp_error(state,
                   "The argument to '-r' must be greater than zero; got %d.",
                   arguments->retries);
      break;
    case 'R':
      arguments->prefix = arg;
      break;
    case 'd':
      arguments->with_debuginfo = arg;
      break;
    case ULP_OP_REVERT_ALL:
      arguments->library = get_basename(arg);
      break;
    case ULP_OP_REVERT:
      arguments->revert = 1;
      break;
    case ULP_OP_DISABLE_THREADING:
      arguments->disable_threads = 1;
      break;

    case ULP_OP_RECURSIVE:
      arguments->recursive = 1;
      break;

    case ULP_OP_DISABLE_SUMMARIZATION:
      arguments->no_summarization = 1;
      break;

    case ULP_OP_ONLY_LIVEPATCHED:
      arguments->only_livepatched = 1;
      break;

    case 'u':
      arguments->user_wildcard = arg;
      break;

#if defined ENABLE_STACK_CHECK && ENABLE_STACK_CHECK
    case 'c':
      arguments->check_stack = 1;
      break;
#endif
    case ARGP_KEY_ARG:
      if (state->arg_num == 0) {
        arguments->command = command_from_string(arg);
        if (arguments->command == ULP_NONE)
          argp_error(state, "Invalid command: %s.", arg);
      }
      else if (state->arg_num <= ARGS_MAX)
        arguments->args[state->arg_num - 1] = arg;
      else
        argp_error(state, "Too many arguments.");
      break;

    case ARGP_KEY_END:
      handle_end_of_arguments(state);
      break;

    case ULP_OP_COLOR:
      if (strcmp(arg, "no") == 0 || strcmp(arg, "n") == 0) {
        no_color = true;
      }
      else if (strcmp(arg, "auto") == 0) {
        no_color = !check_color_available();
      }
      else {
        no_color = false;
      }
      break;

    case ULP_OP_TIMEOUT:
      if (isnumber(arg)) {
        long num = atol(arg);
        if (num < 0)
          argp_error(state, "--timeout value must be a non-negative integer");
        set_run_and_redirect_timeout(num);
      }
      else {
        argp_error(state, "--timeout value must be a non-negative integer");
      }
      break;

    default:
      return ARGP_ERR_UNKNOWN;
  }

  return 0;
}

static bool
check_color_available(void)
{
  /* Check if NO_COLOR env variable is set.  */
  if (getenv("NO_COLOR"))
    return false;

  /* Check if terminal supports ANSI escapes.  */
  if (isatty(STDOUT_FILENO)) {
    const char *term = getenv("TERM");
    if (term && strcmp(term, "dumb")) {
      return true;
    }
  }

  /* ANSI escapes not available.  */
  return false;
}

void
change_color(const char *ansi_escape)
{
  /* If no_color is set, doens't push colors.  */
  if (!no_color) {
    fputs(ansi_escape, stdout);
  }
}

static bool
requires_ptrace(command_t command)
{
  switch(command) {
    case ULP_NONE:
    case ULP_DUMP:
    case ULP_POST:
    case ULP_EXTRACT:
    case ULP_PACKER:
    case ULP_LIVEPATCHABLE:
      return false;

    case ULP_PATCHES:
    case ULP_TRIGGER:
    case ULP_CHECK:
    case ULP_MESSAGES:
    case ULP_SET_PATCHABLE:
      return true;
  }
  return false;
}

int
main(int argc, char **argv, char *envp[] __attribute__((unused)))
{
#ifdef ENABLE_AFL
  while (*envp) {
    const char *key = strtok(*envp, "=");
    const char *val = strtok(NULL, "=");
    if (!strcmp(key, "ULP_IN_AFL_TEST") && val && *val == '1') {
      argv = afl_init_argv(&argc);
      break;
    }
    envp++;
  }
#endif /* ENABLE_AFL  */

  struct arguments arguments = { 0 };
  int ret = 0;

  /* Set no_color here, as user may not have passed --color.  */
  no_color = !check_color_available();

  /* Initialize retries correctly.  */
  arguments.retries = 1;

  argp_parse(&argp, argc, argv, 0, 0, &arguments);


  /* Check if command requires ptrace.  */
  if (requires_ptrace(arguments.command) &&
        check_ptrace_scope() == false) {
    WARN("System has 'ptrace_scope' enabled. Please become root or disable it "
         "by setting:\n\n"
         "$ echo 0 | sudo tee /proc/sys/kernel/yama/ptrace_scope\n\n"
         "and try again.");
    return EPERM;
  }

  switch (arguments.command) {
    case ULP_NONE:
      ret = 1;
      break;

    case ULP_PATCHES:
      ret = run_patches(&arguments);
      break;

    case ULP_CHECK:
      ret = run_check(&arguments);
      break;

    case ULP_DUMP:
      ret = run_dump(&arguments);
      break;

    case ULP_PACKER:
      ret = run_packer(&arguments);
      break;

    case ULP_TRIGGER:
      ret = run_trigger(&arguments);
      break;

    case ULP_POST:
      ret = run_post(&arguments);
      break;

    case ULP_MESSAGES:
      ret = run_messages(&arguments);
      break;

    case ULP_LIVEPATCHABLE:
      ret = run_livepatchable(&arguments);
      break;

    case ULP_EXTRACT:
      ret = run_extract(&arguments);
      break;

    case ULP_SET_PATCHABLE:
      ret = run_set_patchable(&arguments);
      break;
  }

  return ret;
}
